README

=pod

=head1 README

You can use perldoc to read this document: C<perldoc README>.

=head1 INSTALLATION

Prior to running the server/client you must install it. This document gives you
a very broad overview of the neccessary steps, for detailed instructions refer
please to the L<INSTALL> file (f.i. via perldoc INSTALL).

=head2 Verify the package

Before you do anything with the downloaded package, verify the signature.

Download our GnuPG key from http://www.bsi.bund.de/produkte/dicop/download.htm
and then check the detachd signature with:

	gpg --verify Dicop-Proxy-3.00_11.tar.gz.asc

(replace 3.00_11 with the actual version you got)

If the signature does not verify ok, please notify us immidiately. B<DO NOT>
run perl Makefile.PL or anythign else nor install the package when the
signature cannot be verified ok. B<Your system might get compromised!>.

=head2 Requirements

You will need Perl v5.8.3 or better, preferable Perl v5.8.6, to run the proxy.

You also need an HTTP or FTP server to serve the clients with files. The proxy
will answer file requests from the clients with the URL of the file server.

This can either be the same file server that the main server refers to, or
a different one. But it absolutely needs to enable the client to download
the same requested files.

=over 2

=item proxy

When the proxy-daemon is running, you can connect to it via a web browser by
using http://127.0.0.1:8889/ (the default port) and it should return the main
page.

=back

If the proxy is running at another machine, simple connect to it
like you would connect to any other web browser, by using the correct IP and
port.

You can use the perl documentation utility perldoc to get more information. The
documentation is located in the subdirectory C<doc/>.


=head2 Hardware

The maschine to run the proxy should have at least 32 MB (no X-Windows) or
64 MB (with X-Windows) so that no swapping occurs. A 200 Mhz x86 CPU does just
fine for a couple hundred clients, although having one or more Ghz and a
GigaByte or two RAM never hurt ;)


=head2 Supported platforms

The proxy was successfully under the following platforms:

Linux SuSE 9.1
--------------
 Perl v5.8.5		test-environment

Linux SuSE 7.3
--------------
 Perl v5.8.3		development

It should work on Unix/Linux platforms, and even on Mac-OS. However, we can not
guaranty that it will work. If you experience any problems, please contact us
(see below).

=head1 AUTHOR

(C) Bundesamt fuer Sicherheit in der Informationstechnik 1998-2006.

For licensing information please refer to the LICENSE file.

=head2 Contact

	Address: BSI
		 Referat 123
		 Godesberger Alle 185-189
		 Bonn
		 53175
		 Germany
	email:	 dicop@bsi.bund.de		(for public key see dicop.asc)
	www:	 http://www.bsi.bund.de/

Please file bug reports under L<http://rt.cpan.org>.

=cut


CHANGES

=pod

=head1 History revision for DiCoP

This document details the complete changes from v3.01 onwards.

v3.00 was seperated from Dicop-Server and is the first real, working version.

=head2 2006-06-01 (Proxy v3.03)

No functional changes:

=over 2

=item fixed

	removed unused uptime() method
	POD fixes, add POD tests
	increase POD coverage to 100%
	wcr lib: Total: lines 671, words 2215, chars 16674
	43 tests

=back

=head2 2006-05-02 (Proxy v3.02)

No functional changes:

=over 2

=item fixed

	Makefile.PL: use YAML to generate better META.yml
		     use PREREQ_FATAL
		     need Linux::Cpuinfo are prereq
	fix (c) dates, contact address etc
	make a non-beta version for release
	add a SIGNATURE again (for CPAN release)
	wcr lib: Total: lines 685, words 2225, chars 16848
	33 tests

=back

=head2 2005-06-22 (Proxy v3.01 build 3)

=over 2

=item added

	Trouble.pod: explain /etc/protocols and chroot environments
	setup - check and setup a new proxy installation

=item fixed

	amend NEW
	wcr lib: Total: lines 685, words 2225, chars 16848
	33 tests ()

=back

=head2 2005-06-20 (Proxy v3.01 build 2)

=over 2

=item fixed

	adjust requirements in Makefile.PL:
	  Dicop-Base minimum build 3 
	  Math-BigInt v1.77
	  Net-Server v0.86
	version-depended test
	default config file config/proxy.cfg.sample
	wcr lib: Total: lines 685, words 2225, chars 16848
	33 tests ()

=back

=head2 2005-02-16 (Proxy v3.01 build 1)

=over 2

=item fixed

	wcr lib: Total: lines 685, words 2225, chars 16849
	33 tests ()

=back

=head1 AUTHOR

(c) Bundesamt fuer Sicherheit in der Informationstechnik 1998-2005

DiCoP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License version 2 as published by the Free
Software Foundation.

See the file LICENSE or L<http://www.bsi.bund.de/> for more information.

=cut


t/common.pl

use Dicop::Event;
use Dicop::Item;
use Dicop::Request;
use Dicop::Request::Pattern;

###########################################################################
# load request.def file

sub _load_patterns
  {
  my $patterns = [ Dicop::Item::from_file ( "../def/request.def", 'Dicop::Request::Pattern',) ];

  foreach my $pattern (@$patterns)
    {
    $pattern->_construct();
    # check for errors
    if ($pattern->error() ne '')
      {
      request Carp; croak($pattern->error());
      }
    }
  $patterns;
  }

#sub _load_templates
#  {
#  my $file = "../def/objects.def";
#  Dicop::Item::_load_templates( $file );
#  }

# load messages and request patterns
Dicop::Event::load_messages("../msg/messages.txt");

$Dicop::Request::DEFAULT_PATTERNS = _load_patterns();

#_load_templates();

1;


Makefile.PL
use ExtUtils::MakeMaker;
require 5.008003;	# we need at least Perl v5.8.3
use strict;
use lib 'lib';          # if we bundled some modules with the
                        # distribution, try to find these first
                        # (otherwise old core modules would crop up)

my $MIN_BASE_BUILD = 0;	# for CPAN releases, wo always have BUILD == 0

sub MY::metafile {
  package MY;
  my $self = shift;
  if (!eval { require YAML; 1 })
  {
    warn ("YAML not installed, will not override META.yml");
    return $self->SUPER::metafile_target(@_);
  };

  require YAML::Node;
  my $node = YAML::Node->new({});

  $node->{name}                 = $self->{DISTNAME};
  $node->{version}              = $self->{VERSION};
  $node->{version_from}         = $self->{VERSION_FROM};
  $node->{license}              = 'gpl';
  $node->{distribution_type}    = 'module';
  $node->{generated_by}         = "$self->{DISTNAME} version $self->{VERSION}";
  $node->{installdirs}          = 'site';
  $node->{requires}             = $self->{PREREQ_PM};

  my $dump = YAML::Dump( $node );

  $dump =~ s/^(.*)$/\t\$(NOECHO) \$(ECHO) "$1" >>META.yml/gm;
  $dump =~ s/>>META\.yml/>META.yml/;

  return "metafile:\n$dump";
}

# for VERSION and BUILD
eval { require Dicop::Proxy; };

if (!defined $Dicop::Proxy::VERSION)
  {
  warn ("\n*********** Error:\n\n");
  die ("  Can't load Dicop::Proxy - make sure you installed Dicop::Base first\n\n");
  }

# SSL does not work atm, so don't bother to check for it:

## check that we have SSL support:
#my $ssl = 0;
#eval { require IO::Socket::SSL; };
#if (defined $IO::Socket::SSL::VERSION)
#  {
#  $ssl++;
#  $ssl ++ if $IO::Socket::SSL::VERSION > 0.95;
#  }
#
#if ($ssl < 2)
#  {
#  warn ("*********** Warning:\n");
#  warn ("  You do not seem to have IO::Socket::SSL installed. SSL support will not be available.\n")
#    if $ssl == 0;
#  warn ("  Need at least IO::Socket::SSL v0.95 for proper SSL support, but found only v$IO::Socket::SSL::VERSION.\n")
#    if $ssl == 1;
#  warn ("  You need to install the latest version of IO::Socket::SLL from http://search.cpan.org/\n");
#  sleep(5);
#  }

use Dicop::Proxy;			# for VERSION and BUILD
my $version = $Dicop::Proxy::VERSION;
# for releases, omit the BETA VERSION
$version .= "_$Dicop::Proxy::BUILD" if $Dicop::Proxy::BUILD > 0;

WriteMakefile(
    'NAME'		=> 'Dicop-Proxy',
    'VERSION'		=> $version,
    'PREREQ_PM'		=> { 
		'Dicop::Base' => 3.02,
		'Digest::MD5' => 2.09,
		'HTML::Tagset' => 3.03,
		'HTTP::Request' => 1.24,
		'Math::BigInt' => 1.77,
		'Net::Server' => 0.86,
		'Time::HiRes' => 1.2,
		'File::Spec' => 0.82,
		'Test::Simple' => 0.47,
		'Linux::Cpuinfo' => 1.7,
    },
    'PREREQ_FATAL' => 1,
);

warn ("\n **** Warning: Need at least Dicop::Base build $MIN_BASE_BUILD, but got only $Dicop::Base::BUILD! ****\n\n")
  unless $Dicop::Base::BUILD >= $MIN_BASE_BUILD;



lib/Dicop/Proxy/Config.pm
#############################################################################
# Dicop::Proxy::Config - define valid config keys
#
# (c) Bundesamt fuer Sicherheit in der Informationstechnik 2003-2006
#
# DiCoP is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License version 2 as published by the Free
# Software Foundation.
#
# See the file LICENSE or L<http://www.bsi.de/> for more information.
#############################################################################

package Dicop::Proxy::Config;
use vars qw($VERSION);
$VERSION = 0.02;	# Current version of this package
require  5.008003;	# requires this Perl version or later

use strict;

sub allowed_keys
  {
  # setup the allowed keys and their type

  my $ALLOWED_KEYS = {};
  foreach my $key (qw/ 
   hand_out_work background
   /)
    { $ALLOWED_KEYS->{$key} = 'flag'; }

  foreach my $key (qw/ 
   min_chunk_size
   max_chunk_size
   /)
    { $ALLOWED_KEYS->{$key} = 'minutes'; }

  foreach my $key (qw/ 
   max_request_time
   initial_sleep
   /)
    { $ALLOWED_KEYS->{$key} = 'seconds'; }


  #foreach my $key (qw/ 
  # /)
  #  { $ALLOWED_KEYS->{$key} = 'hours'; }

  foreach my $key (qw/ 
   name
   group
   user
   host
   proto
   chroot
   title
   allow_admin allow_status allow_stats allow_work
   deny_admin deny_status deny_stats deny_work
   error_log
   server_log
   log_level
   client_architectures
   msg_file
   patterns_file
   objects_def_file
   /)
    { $ALLOWED_KEYS->{$key} = 'string'; }
  
  foreach my $key (qw/ 
   mail_admin
   mail_from
   mail_to
   mail_errors
   /)
    { $ALLOWED_KEYS->{$key} = 'email'; }
  
  foreach my $key (qw/ 
   self
   file_server
   server
   /)
    { $ALLOWED_KEYS->{$key} = 'URL'; }
   
#mail_server

  foreach my $key (qw/ 
   port
   require_client_build
   max_requests
   id
   /)
    { $ALLOWED_KEYS->{$key} = 'int'; }
  
  foreach my $key (qw/ 
   require_client_version
   /)
    { $ALLOWED_KEYS->{$key} = 'float'; }
  
  foreach my $key (qw/ 
   log_dir def_dir msg_dir tpl_dir mailtxt_dir
   /)
    { $ALLOWED_KEYS->{$key} = 'dir'; }

  $ALLOWED_KEYS;
  }

1; 

__END__

#############################################################################

=pod

=head1 NAME

Dicop::Proxy::Config - define valid config keys for a Dicop-Proxy

=head1 SYNOPSIS

	use Dicop::Proxy::Config;
	use Dicop::Config;

	my $config = Dicop::Config->new(
		'data/server.cfg', 
		Dicop::Proxy::Config::allowed_keys() );

=head1 REQUIRES

perl5.008003

=head1 EXPORTS

Exports nothing.

=head1 DESCRIPTION

This module contains a method to create the list of allowed config keys.

=head1 METHODS

=head2 allowed_keys()

	$allowed_keys = Dicop::Proxy::Config::allowed_keys();

=head1 BUGS

None known yet.

=head1 AUTHOR

(c) Bundesamt fuer Sicherheit in der Informationstechnik 2003-2006

DiCoP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License version 2 as published by the Free
Software Foundation.

See L<http://www.bsi.de/> for more information.

=cut



lib/Dicop/Proxy.pm
#############################################################################
# Dicop::Proxy - a proxy for Dicop-Server
#
# (c) Bundesamt fuer Sicherheit in der Informationstechnik 1998-2006
#
# DiCoP is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License version 2 as published by the Free
# Software Foundation.
#
# See the file LICENSE or L<http://www.bsi.de/> for more information.
#############################################################################

package Dicop::Proxy;
use vars qw($VERSION $BUILD);
$VERSION = '3.03';	# Current version of this package
$BUILD = 0;		# Current build of this package
require  5.008003;	# requires this Perl version or later

use base qw/Dicop::Handler/;

use strict;
use Dicop::Cache;
use Dicop::Event qw/give_up/;
use Dicop::Proxy::Config;
use Dicop::Base;

use Dicop::Connect qw/_connect_server _load_connector/;

# fields used by the proxy part:
#  _info_cache	: contains for each client that connected since the last
#		  time we talked to the server an info-request, sorted by id
#  _send_cache	: contains all requests (minus info requests) that need to be
#		  send to the server
#  _send_id	: last request ID in send cache
#  _req_remap	: maps incoming client request numbers to the ones we sent to
#		  the server to remap server answers back to the client

#############################################################################
# Data flow is:

# For each connect, pre_connect() is called once before handle_requests(). The
# latter will call parse_requests() to create a list of requests. For each
# request, one of:

# request_file
# request_work
# request_test 
# status_main | status_help | status_config | status_style

# will be called. These will either cache some requests to be sent to the
# server, or generate the different status pages (these all land in status()
# in Handler.pm and are handled generic). After all requests (typically only
# one anyway) have been handled that way, finish_connect() will be called and
# if there were any requests to be send to the server, will send them and then
# relay back the answer to the client.

#############################################################################

sub type () { 'proxy'; }

# check_auth_request needs to return a fake client object to signal that
# everything went ok. We return ourselves, and thus we need is_proxy():
sub is_proxy { 0; }

sub version { $VERSION; }
sub build { $BUILD; }

sub _init
  {
  # init my own fields, than call base class _init() to finish it off
  my ($self,$args) = @_;
  
  $self->{_send_cache} = {};
  $self->{_info_cache} = {};
  $self->{_send_id} = 'req0001';
  $self->{_req_remap} = {};
  $self->{debug} = 0;

  # disable checks for files, because they don't exist at the proxy
  $self->{no_file_check} = 1;

  # XXX TODO: we fake ourself as client object
  $self->{name} = 'unknown';

  # we cache test request answers for a maximum of 6 hours
  $self->{_test_cache} = Dicop::Cache->new( timeout => 3600*6 );

  $self->SUPER::_init($args);

  ###########################################################################  
  # set up connector (.e.g. the method we are using to connect server)

  my $cfg = $self->{config};

  $self->_load_connector($cfg, $args);

  my $pid = $$ || 0; my $arch = $^O;

  # get CPU info
  my $cpuinfo = Dicop::Base::cpuinfo($self, $args->{_no_warn});

  # build our auth request that we will use to authenticate ourself to the main server
  $self->{ident} = Dicop::Request->new( id => 'req0001',
   data => "cmd_auth;id_$cfg->{id};version_$VERSION-$BUILD;pid_$pid;arch_$arch$cpuinfo",
   patterns => $self->{request_patterns},
   );

  # init done
  $self;
  }

sub _after_config_read
  {
  # check config for being valid and insert default values
  my ($self,$cfg,$cfgdir,$cfgfile) = @_;

  ###########################################################################
  # set some default config values

  Dicop::Base::cfg_default( $self,
    msg_dir => 'msg',
    def_dir => 'def',
    tpl_dir => 'tpl',
    max_requests => 128,
    msg_file => 'messages.txt',
    patterns_file => 'request.def',
    log_dir => 'logs',
    error_log => 'error_proxy.log',
    server_log => 'proxy.log',
    file_server => 'http://127.0.01/',
    max_request_time => 5,
    self => '/',
    client_architectures => 'linux,os/2,mswin32,armv4l',
    chroot => '',
    name => "dicop-proxy",
    title => "Dicop-Proxy - ",
    );

# mail_server => 'none',

  $self;
  }

sub _load_data
  {
  my $self = shift;

  my $cfg = $self->{config};
  # basic check for keys/values/types
  my $allowed_keys = Dicop::Proxy::Config::allowed_keys();
  my $check = $cfg->check($allowed_keys);

  give_up($check) if defined $check;
  
  $self->{allowed_archs} = $cfg->{client_architectures}
    || give_up ('Need a list of allowed client architectures');
  $self->{allowed_archs} = [ split(/\s*,\s*/, $self->{allowed_archs}) ];

  }

sub pre_connect
  {
  my ($self,$peer,$params) = @_;

  # XXX TODO: this might cause problems with proxy-chaining
  $self->{peeraddress} = $peer;

  # for each connect, reset the send_id etc
  $self->clear_cache();
  }

sub finish_connect
  {
  my ($self,$result) = @_;

  $self->output ("finish_server\n");

  # if $$result is non empty, some error occured or a status page was
  # requested and successfully build
  return $result if $$result ne '';
    
  # if it is empty, send our accumulated requests to the server and
  # return either the response or some error message to the client
  $$result = $self->talk_to_server();

  $result;
  }

#############################################################################

sub check_auth_request
  {
  # check the auth or info request a client/proxy sent us for basic validity
  # Return ref to $client, or error message
  my ($self,$req,$rid,$check_proxy) = @_;

  # XXX TODO: auth request from client should be checked for basic validity
  # How? We do not know which clients are valid. Define valid client range
  # in config file? Check architecture to be known?

  # XXX TODO: In addtion, we can check for the architecture being valid.

  # cheap hack to have a valid client object
  $self;
  }

sub request_auth
  {
  my ($self,$auth,$info) = @_;

  # XXX TODO: auth request from client should be checked for basic validity
  # How? We do not know which clients are valid. Define valid client range
  # in config file? Check architecture to be known?

  return ( {}, '', {});
  }

#############################################################################

sub remap_request_id
  {
  # map from server's response request_id back to what the client sent us
  my ($self,$req_id) = @_;

  $self->{_req_remap}->{$req_id};
  }

sub cache_request
  {
  # takes a request, and adds it to the _send_cache
  # return 1 if we need to send to the server, otherwise 0
  my ($self,$request,$client,$info) = @_;

  # create a new request with a new id for the cache
  my $req = $request->copy();
  my $old_id = $request->request_id();
  my $new_id = ++$self->{_send_id};
  $req->request_id($new_id);
  $self->{_send_cache}->{$new_id} = $req;

  # enter this request into the remap cache so that we can later remap the
  # server's answer back to the client:
  $self->{_req_remap}->{$new_id} = $old_id;

  # in addition to the cached request, we also need to create an info request

  # go through all items in the info cache and if the cached request
  my $info_req;					# the info request
  my $ic = $self->{_info_cache};
  for my $id (keys %$ic)
    {
    my $req = $ic->{$id};
  
    # found the right info request? so remember it and stop 
    $info_req = $req, last if ($req->{id} eq $info->{id});
    }

  if (defined $info_req)
    {
    # we already have one, so simple add the current request to the 'for' field
    $info_req->{for} .= ",$self->{_send_id}";	# this info is for this request
    }
  else
    {
    # don't have one yet, so create a new one
    $info_req = $info->copy();			# copy the auth from client
    $info_req->{cmd} = 'info';			# turn into info request
    $info_req->{for} = $self->{_send_id};	# this info is for this request
    $info_req->{ip} = $self->{peeraddress};	# client's real IP
    $info_req->request_id(++$self->{_send_id});	# give it a new request id
  
    # store the info request in our cache
    $self->{_info_cache}->{$info_req->request_id()} = $info_req;
    }

  1;						# XXX TODO: rc not used anymore
  }

sub build_server_talk
  {
  # create the text that is sent to the server
  my $self = shift;

  # string together request and info requests
  my $txt = $self->{ident}->as_request_string() . "&";
  foreach my $id (sort keys %{$self->{_send_cache}})
    {
    my $req = $self->{_send_cache}->{$id};
    print STDERR "error in request $id: ", $req->error() if $req->error();
    $txt .= ($req->as_request_string()||'') . "&";
    }
  foreach my $id (sort keys %{$self->{_info_cache}})
    {
    my $req = $self->{_info_cache}->{$id};
    print STDERR "error in request $id: ", $req->error() if $req->error();
    $txt .= ($req->as_request_string()||'') . "&";
    }
  $txt =~ s/\&$//;		# remove last &
  $txt;
  }

sub clear_cache
  {
  my $self = shift;

  $self->{_send_cache} = {};
  $self->{_info_cache} = {};
  $self->{_send_id} = 'req0001'; 	# reset id for next send
  $self->{_req_remap} = {};
  $self;
  }
   
sub talk_to_server
  {
  # send anything in cache and the info requests to the server
  my ($self) = @_;
 
  my $txt = $self->build_server_talk();

  my $res = $self->_connect_server(undef,$txt);

  my $response = "<PRE>\nreq0000 099 Proxy localtime ".localtime()."\n";

  if ($res->is_success())
    {
    $self->output($res->content) if $self->{debug} > 1;
    $response .= $res->content;

    # remap now server's responses back to client (any unknown request id is
    # passed through)
    $response =~ s/\b(req\d\d\d\d)(\s+\d)/ ($self->remap_request_id($1) || $1) . $2; /eg;
    }
  else
    {
    # some error occured, send error message to client
    $response .= "req0000 500 Error while talking to upstream server: " . $res->code() . " " . $res->message()."\n";
    }

  # now send what we got from the server back to the client
  $response;
  }

sub output
  {
  my $self = shift;

  return unless $self->{debug} > 0;

  print STDERR join (" ", @_),"\n";
  }

sub _sleeping
  {
  # called by _connect_server()
  return;
  }

sub _die_hard
  {
  # called by _connect_server()
  return;
  }

#############################################################################
# the actual client requests/reports land us here

sub request_test
  {
  my ($self,$request,$client,$info) = @_;

  $self->cache_request($request,$client,$info);
  
  \"";
  }

sub request_work
  {
  # request test case (take it either form cache or ask server)
  my ($self,$request,$client,$info) = @_;

  $self->cache_request($request,$client,$info);

  \"";
  }

sub report
  {
  # client want's to report work or test case result
  my ($self,$request,$client,$info) = @_;

  $self->cache_request($request,$client,$info);

  \"";
  }

sub _client_from_info
  {
  # Not used here, called by Base::Handler::request_file()
  # XXX TODO: this would break if a proxy (instead of a client) connects to us?
  my ($req,$client,$info) = @_;

  (undef,$client,undef);
  }

1; 

__END__

#############################################################################

=pod

=head1 NAME

Dicop::Proxy - A proxy for a Dicop system

=head1 SYNOPSIS

	use Dicop::Proxy;

	$proxy = Dicop::Proxy->new();
	$proxy->parse_client_requests($self,$requests);

=head1 DESCRIPTION

This module includes some routines that make up the logic of the proxy part
of the server.

This is a subclass of Dicop::Handler and overrides only the necc. methods.

This package is used by C<dicopp> to implement a full Dicop-Proxy.

=head1 REQUIRES

perl5.008003, Exporter

=head1 EXPORTS

Exports nothing per default.

=head1 METHODS

=head2 remap_request_id()
  
	my $req_id = $proxy->remap_request_id( $response->request_id() );

Map from server's response request_id back to what the client sent us.

=head2 cache_request()

Takes a request, and adds it to the _send_cache. It also creates a client_info
request, if that isn't already there, otherwise it updates the client info.
Return 1 if we need to send to the server (cache full), otherwise 0.

=head2 build_server_talk()
  
Create the text that is sent to the server by stringing together request and
info requests from the cache.

=head2 clear_cache()

Clear the send cache and the client info list, reset the request id for the
send cache.
  
=head2 new()

Create a new Dicop::Proxy object and init the data fields of it. Called by the
servers init() method.

=head2 talk_to_server()

Uses L<build_server_talk> to build the text for the upstream server, send it
to the server, and then clear the cache and client info list. Returns the
servers answer as text.

=head2 request_test()

Called for each testcase request. 

=head2 request_work()

Called for each work request. 

=head2 request_file()

Called for each file request. 

=head2 report()

Called for each chunk the client reports back.

=head2 type()

Returns the string 'proxy'.

=head2 is_proxy()

Returns true.

=head2 output()

Output a string to STDERR for debugging purposes.

=head1 BUGS

None known yet.

=head1 AUTHOR

(c) Bundesamt fuer Sicherheit in der Informationstechnik 1998-2006

DiCoP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License version 2 as published by the Free
Software Foundation.

See L<http://www.bsi.de/> for more information.

=cut

